File: cBooks/scripts/book.py
#!/usr/bin/python3
#!/usr/bin/env python3
import locale
import os
import sys
import locale
import os
import sys

import cgi
import os
import json
import sys
import base64 as b64
from chinese import *
from modules import *
print("Content-type: text/html\n\n")
print("")
form = cgi.FieldStorage()
Action = form.getfirst("RequestType")
JSONbuf = {}
jEncode = json.JSONEncoder().encode
bookDir = WEB_ROOT+"cBooks/books/"
SubDivider = Chin["interpunct"]   # symbol to indicate subdivision by category ex 賦-卷一.html
Entry=Creator(FindRegex,rgx='(<div class="biji_block">.*?</div>)(?=\W{0,}<div class="biji_block">|\W{0,}</body>)')

if Action == "ListBooks":
    JSONbuf['booklist'] = shell_exec("ls {0}".format(bookDir)).decode("utf-8").split("\n")[:-1]
    
elif Action == "ListChapters":
    targetBook=form.getfirst("BookName")
    targetBook=b64.urlsafe_b64decode(targetBook).decode("utf-8")
    BookRoot="{0}{1}/".format(bookDir,targetBook).replace(";","")
    chapterFiles = os.listdir(BookRoot)
    JSONbuf['chapterlist'] = [x.replace(BookRoot,"").replace("/",SubDivider).split(".")[0] for x in ChiBookSort(chapterFiles)]
        
    
elif Action == "ListChapters":
    targetBook=form.getfirst("BookName")
    targetBook=b64.urlsafe_b64decode(targetBook).decode("utf-8")
    BookRoot="{0}{1}/".format(bookDir,targetBook).replace(";","")
    try:
        for subDir in os.listdir(BookRoot):
            chapterFiles += [SubDivider.join([subDir,x]) for x in os.listdir(BookRoot+subDir)]
    except NotADirectoryError:
        chapterFiles = os.listdir(BookRoot)

    JSONbuf['chapterlist'] = [x.replace(BookRoot,"").replace("/",SubDivider).split(".")[0] for x in ChiBookSort(chapterFiles)]

elif Action == "SearchContent":
    Book=b64.urlsafe_b64decode(form.getfirst("BookName")).decode('utf-8')
    Chapter=b64.urlsafe_b64decode(form.getfirst("ChapterName")).decode('utf-8')
    Contents=b64.urlsafe_b64decode(form.getfirst("SearchContents")).decode('utf-8')
    Range=b64.urlsafe_b64decode(form.getfirst("SearchRange")).decode('utf-8')
    Comment=Creator(FindRegex,rgx='<span class="zhipi">(.*?)</span>',Remove=True if Range=="zhengwen" else False)
    TargetPath = "^.*?\.html" # default search through everything
    if Chapter != "*":    # specific Chapter -> specific Book
        TargetPath = "{0}/{1}.html".format(Book,Chapter)
    elif Book != "*":     # otherwise all Chapters within specific Book
        TargetPath = ("{0}/".format(Book))+TargetPath
    
    TargetPath = bookDir+TargetPath
    S=Node()
    [d+Entry for d in populate_data(S,TargetPath,Recursive=True)]
    [e+Comment for e in Entry.children]
    TargetData = Comment if Range == "zhushi" else Entry
    JSONbuf.__setitem__("results",[])
    for e in TargetData==Contents:
        e.content = re.sub("({0})".format(Contents),'<span class="match_text">{0}</span>'.format(r"\1"),e.content.decode("utf-8"),0,re.DOTALL|re.MULTILINE)
        parentFile=(-(e<<File))
        cur_book = parentFile.split("/")[-2:]
        cur_book[0] = Chin["open_title_caret"]+cur_book[0]+Chin["close_title_caret"]
        FinalSource = bytes(Chin["interpunct"].join(cur_book).replace(".html",""),encoding="utf-8")
        JSONbuf['results'].append({'source':b64.urlsafe_b64encode(FinalSource).decode("utf-8"),'contents':b64.urlsafe_b64encode(bytes(-e,encoding="utf-8")).decode("utf-8")})
print(jEncode(JSONbuf))

File: cBooks/scripts/book.js


var BaseScript="/cBooks/scripts/book.py?RequestType=";
var HTMLbegin = '<html><head><meta charset="utf-8"><link rel="stylesheet" type="text/css" href="/cBooks/html/chinese.css"></head><body>';
var HTMLend = '</body></html>';

var IS_IOS = IsIOS();

var CommentField = '<span class="comment_field"><input type="text" size="40" id="comment_content" value=""/><button type="button" id="add_button_field">search</button></span>';
var iPadComment;

function AddComment(e)
{
    clearTimeout(iPadComment);
    var BookName    = Base64.encode($("#book_name option:selected").text());
    var ChapterName = Base64.encode($("#book_chapter option:selected").text());
    var Selected = window.getSelection();
    var ParentText = Selected.anchorNode.textContent;
    var BijiBlock = $("div.biji_block:contains('"+ParentText+"')").find(".biji_head").clone();
    if (BijiBlock.text() == "")
    {
	return;
    }
    //BijiBlock.find("span").remove();
    var BijiTitle = BijiBlock.html();
    var SelectedText = ParentText.substring(Selected.anchorOffset-10,Selected.anchorOffset);
    
    if (IS_IOS)
    {
	SelectedText += ParentText.substring(Selected.extentOffset+1,Selected.extentOffset); // in iOS context, usually will want to add comments after a stop mark which is often to the immediate right of the selection 	
    }
    if (SelectedText.length > 0 && ParentText.length > 0)
    {
	var Comment='';
	if (! IS_IOS)
	{
	    Comment=prompt("Insert a comment after "+SelectedText);	    
	    if (Comment.length == 0)
	    {
		return
	    }
	    Comment=Base64.encode(Comment);
	    BijiTitle=Base64.encode(BijiTitle);
	    SelectedText=Base64.encode(SelectedText);
	    var FinalURI = BaseScript+"AddComment&BookName="+BookName+"&ChapterName="+ChapterName+"&BlockName="+BijiTitle+"&SelectedText="+SelectedText+"&CommentText="+Comment;
	    $.getJSON(FinalURI);
	}
	else
	{
	    iPadComment = setTimeout(function(){
		Comment=prompt("Insert a comment after "+SelectedText);	    
		if (Comment.length == 0)
		{
		    return;
		}
		Comment=Base64.encode(Comment);
		BijiTitle=Base64.encode(BijiTitle);
		SelectedText=Base64.encode(SelectedText);
		var FinalURI = BaseScript+"AddComment&BookName="+BookName+"&ChapterName="+ChapterName+"&BlockName="+BijiTitle+"&SelectedText="+SelectedText+"&CommentText="+Comment;
		$.getJSON(FinalURI);		
	    },5000);
	}
    }
}
function ToggleCommentary()
{
    if ($(".zhipi").css('display') == 'inline')
    {
	$(".zhipi").css('display','none');
    }
    else
    {
	$(".zhipi").css('display','inline');
    }
}
function SearchContent()
{
    
    var BookName    = Base64.encode($("#book_name option:selected").text());
    var ChapterName = Base64.encode($("#book_chapter option:selected").text());
    var SearchContents = Base64.encode($("#search_content").val());
    $("#baiwen").attr('checked',false);

    if (BookName == "" )
    {
	BookName = "Kg==";    // b64-encoded "*" to signify wildcarding
    }
    if (ChapterName == "")
    {
	ChapterName = "Kg==";
    }
    var FinalURI = BaseScript+"SearchContent&BookName="+BookName+"&ChapterName="+ChapterName+"&SearchContents="+SearchContents+"&SearchRange="+Base64.encode($("#search_range").val());
    $.getJSON(FinalURI,function(data)
	      {
		  Dbg(data);
		  var buf = '';
		  for (var i = 0; i < data.results.length; i++)
		  {
		      var Source = Base64.decode(data.results[i].source);
		      var Contents = Base64.decode(data.results[i].contents);
		      if (Contents.search("biji_head") == -1 ) // no biji "head" if you're dealing with a paragraph from a novel.  So use the source as the "head"
		      {
			  Contents = '<div class="biji_head">'+Source+'</div>'+Contents;
		      }
		      else
		      {
			  Contents = Contents.replace(/<div class="biji_head">(.*?)<\/div>/,'<div class="biji_head"><span class="source_note">'+Source+"<br></span>$1"+"</div>")
			  
		      }
		      buf += Contents;
		  }
		  $("#mainContent").html(HTMLbegin+buf+HTMLend);
	      }
	     );
    return;
}

function ShowSearchButton()
{
    Reveal($("#search_button"));
    Reveal($("#search_zhipi_field"));
}

function GetChapterContent(e)
{
    var Chapter = Base64.encode(e.currentTarget.value);
    var Book = Base64.encode($("#book_name option:selected").text());
    var FinalURI = BaseScript+"GetChapter&BookName="+Book+"&ChapterName="+Chapter;
    $.getJSON(FinalURI,function(data)
	      {
		  $("#mainContent").html(Base64.decode(data.chapter_contents));
		  var FirstEntry = $(".biji_block")[0];
		  FirstEntry.innerHTML=CommentField+FirstEntry.innerHTML;
		  if (IS_IOS)
		  {
		      $(".biji_entry").css("font-size","18pt");
		      $(".biji_head").css("font-size","20pt");
		      
		  }
	      }
	     );
    $("#baiwen").attr('checked',false);

}

function GetBookChapters(e)
{
    var Selection=Base64.encode(e.currentTarget.value);
    var FinalURI = BaseScript+"ListChapters&BookName="+Selection;
    $.getJSON(FinalURI,function(data)
	      {
		  var ChapterBuf = $("#book_chapter");
		  var NameBuf = '<option value=""></option>\n';
		  for (var i = 0; i < data.chapterlist.length; i++)
		  {
		      
		      NameBuf += '<option value="'+data.chapterlist[i]+'">'+data.chapterlist[i]+'</option>\n';		      
		  }
		  $("#book_chapter").html(NameBuf);
	      }
	     );
    $("#baiwen").attr('checked',false);

    
}


function FirstTime()
{
    $.getJSON(BaseScript+"ListBooks",function(data)
	      {
//		  Dbg(data);
		  var NameBuf=$("#book_name").html();
		  for (var i = 0; i < data.booklist.length; i++)
		  {
		      NameBuf += '<option value="'+data.booklist[i]+'">'+data.booklist[i]+'</option>\n';
		  }
		  $("#book_name").html(NameBuf);
	      }
	     );
    $("#book_name").change(GetBookChapters);
    $("#book_chapter").change(GetChapterContent);
    $("#search_content").change(ShowSearchButton);
    $("#search_button").click(SearchContent);
    $("#baiwen").click(ToggleCommentary);
    $("#add_comment").click(AddComment);

    if (IS_IOS)
    {
	$(".console").css('top','10pt');
    }
    else
    {
	$("#add_comment").css("display","inline");

    }
}

$(document).ready(FirstTime);
if (IS_IOS)
{
    document.addEventListener("selectionchange",AddComment);
    
}
